File: src/drivers/WebDriver.ts
import { Browser, BrowserContext, Page, chromium, firefox, webkit } from 'playwright';
import type { EnvironmentConfig } from '../../config/types';

export interface WebDriverOptions {
  headless?: boolean;
  slowMo?: number;
}

export class WebDriver {
  private _browser: Browser | null = null;
  private _context: BrowserContext | null = null;
  private _page: Page | null = null;

  constructor(private config: EnvironmentConfig) {}

  async launch(opts?: WebDriverOptions): Promise<void> {
    const browserName = this.config.web.browserName;
    const launcher = { chromium, firefox, webkit }[browserName];
    this._browser = await launcher.launch({
      headless: opts?.headless ?? this.config.web.headless,
      slowMo: opts?.slowMo ?? this.config.web.slowMo
    });
    this._context = await this._browser.newContext({
      baseURL: this.config.web.baseUrl,
      viewport: this.config.web.viewport ?? { width: 1280, height: 720 }
    });
    this._page = await this._context.newPage();
  }

  async close(): Promise<void> {
    await this._page?.close();
    await this._context?.close();
    await this._browser?.close();
    this._page = null;
    this._context = null;
    this._browser = null;
  }

  get page(): Page {
    if (!this._page) throw new Error('Browser not launched. Call launch() first.');
    return this._page;
  }

  get context(): BrowserContext {
    if (!this._context) throw new Error('Browser not launched. Call launch() first.');
    return this._context;
  }

  get browser(): Browser {
    if (!this._browser) throw new Error('Browser not launched. Call launch() first.');
    return this._browser;
  }

  get isLaunched(): boolean {
    return this._browser !== null;
  }

  async takeScreenshot(name: string): Promise<Buffer> {
    return this.page.screenshot({ path: `reports/screenshots/${name}.png`, fullPage: true });
  }
}

File: java/conductor-core/src/main/java/com/nouhouari/conductor/drivers/WebDriver.java
package com.nouhouari.conductor.drivers;

import com.microsoft.playwright.Browser;
import com.microsoft.playwright.BrowserContext;
import com.microsoft.playwright.BrowserType;
import com.microsoft.playwright.Page;
import com.nouhouari.conductor.config.EnvironmentConfig;
import com.nouhouari.conductor.config.Viewport;
import com.nouhouari.conductor.world.PlaywrightHolder;

import java.nio.file.Path;
import java.nio.file.Paths;

/** Java port of src/drivers/WebDriver.ts. */
public class WebDriver {

    private final EnvironmentConfig config;
    private Browser browser;
    private BrowserContext context;
    private Page page;

    public WebDriver(EnvironmentConfig config) {
        this.config = config;
    }

    public void launch() {
        launch(null, null);
    }

    public void launch(Boolean headlessOverride, Integer slowMoOverride) {
        var web = config.web();
        BrowserType launcher = switch (web.browserName()) {
            case "firefox" -> PlaywrightHolder.get().firefox();
            case "webkit" -> PlaywrightHolder.get().webkit();
            default -> PlaywrightHolder.get().chromium();
        };

        BrowserType.LaunchOptions launchOptions = new BrowserType.LaunchOptions()
                .setHeadless(headlessOverride != null ? headlessOverride : web.headless());
        Integer slowMo = slowMoOverride != null ? slowMoOverride : web.slowMo();
        if (slowMo != null) {
            launchOptions.setSlowMo(slowMo.doubleValue());
        }
        browser = launcher.launch(launchOptions);

        Viewport viewport = web.viewport();
        context = browser.newContext(new Browser.NewContextOptions()
                .setBaseURL(web.baseUrl())
                .setViewportSize(viewport != null ? viewport.width() : 1280,
                        viewport != null ? viewport.height() : 720));
        page = context.newPage();
    }

    public void close() {
        if (page != null) {
            page.close();
        }
        if (context != null) {
            context.close();
        }
        if (browser != null) {
            browser.close();
        }
        page = null;
        context = null;
        browser = null;
    }

    public Page getPage() {
        if (page == null) {
            throw new IllegalStateException("Browser not launched. Call launch() first.");
        }
        return page;
    }

    public BrowserContext getContext() {
        if (context == null) {
            throw new IllegalStateException("Browser not launched. Call launch() first.");
        }
        return context;
    }

    public Browser getBrowser() {
        if (browser == null) {
            throw new IllegalStateException("Browser not launched. Call launch() first.");
        }
        return browser;
    }

    public boolean isLaunched() {
        return browser != null;
    }

    public byte[] takeScreenshot(String name) {
        Path path = Paths.get("reports/screenshots", name + ".png");
        return getPage().screenshot(new Page.ScreenshotOptions().setPath(path).setFullPage(true));
    }
}
